### PR TITLE
Updated download page as per DEP-20.

### PR DESCRIPTION
**Pull request by sarahboyce**
See newly approved DEP-20: https://github.com/django/deps/pull/109

New download page:

Screenshot from 2026-08-07 12-48-32

Sample roadmap page:

Screenshot from 2026-08-07 12-52-13

As we may announce with a blog post, it would be nice if this page is updated accordingly

We perhaps want to time the updates alongside the following docs that need updating:
- https://docs.djangoproject.com/en/6.1/faq/install/#faq-python-version-support
- https://docs.djangoproject.com/en/6.1/internals/release-process/

### CODE REVIEW

File: djangoproject/templates/releases/download.html [hunk 1]
```diff
@@ -95,19 +100,19 @@ <h2 id="supported-versions">Supported Versions</h2>
...
+      <td>{% get_latest_release_version '6.0' %}</td>
       <td>August 4, 2026</td>
       <td>April 2027</td>
     </tr>
     <tr>
       <td>6.1</td>
-      <td>{% get_latest_micro_release '6.1' %}</td>
+      <td>{% get_latest_release_version '6.1' %}</td>
       <td>April 2027</td>
       <td>December 2027</td>
```
Comment (jacobtylerwalls) [inline djangoproject/templates/releases/download.html:116]: We might decide to bump this to January 2028, as otherwise for a single month period there will be no version in mainstream support.

File: releases/management/commands/generate_release_roadmap.py [hunk 2]
```diff
@@ -64,24 +53,65 @@
...
+    },
+    {
+        "name": "6.2",
+        "is_lts": True,
+        "release_date": dtime.date(2027, 4, 1),
+        "mainstream_end": dtime.date(2027, 12, 1),
+        "extended_end": dtime.date(2030, 4, 1),
+    },
+    {
+        "name": "2028.0",
```
Comment (jacobtylerwalls) [inline releases/management/commands/generate_release_roadmap.py]: I would generate the SVG with just "2028", since 2028.0 is only good for about 30 days.

File: djangoproject/templates/releases/download.html [hunk 3]
```diff
@@ -70,16 +70,21 @@ <h2>After you get it</h2>
...
   <h2 id="supported-versions">Supported Versions</h2>
-  <p><strong>Feature releases</strong> (A.B, A.B+1, etc.) will happen roughly every eight months.
-    These releases will contain new features, improvements to existing features, and such.</p>
-  <p><strong>Patch releases</strong> (A.B.C, etc.) will be issued as needed, to
-    fix bugs and/or security issues. These releases will be 100% compatible with
+  <p><strong>Feature releases</strong> contain new features, improvements to existing
+    features, and other enhancements. Beginning with Django 2028.0, feature releases
+    will use the version format YYYY.0 and will be released every January. Earlier
+    feature releases used the version format A.B and were released roughly every eight
+    months.</p>
```
Comment (jacobtylerwalls) [inline djangoproject/templates/releases/download.html]: A.B is analogous to YYYY, I think, not YYYY.N

File: djangoproject/templates/releases/download.html [hunk 4]
```diff
@@ -70,16 +70,21 @@ <h2>After you get it</h2>
...
+    needed, to fix bugs and/or security issues. These releases are 100% compatible with
     the associated feature release, unless this is impossible for security
     reasons or to prevent data loss. So the answer to "should I upgrade to the
     latest patch release?” will always be "yes."</p>
-  <p>Certain feature releases will be designated as <strong>long-term support
-    (LTS) releases</strong>. These releases will get security and data loss
-    fixes applied for a guaranteed period of time, typically three years.</p>
+  <p>Django 6.2 is the final release under the previous versioning and support policy.
+    Under that policy, only designated long-term support (LTS) releases received the
+    full three-year support period. Starting with Django 2028.0, every feature release
```
Comment (jacobtylerwalls) [inline djangoproject/templates/releases/download.html]: 2028 here as well?

File: djangoproject/templates/releases/download.html [hunk 5]
```diff
@@ -70,16 +70,21 @@ <h2>After you get it</h2>
...
+    receives the same three-year support period. Refer to the tables below for the
+    support periods of individual release series.</p>
```
Comment (jacobtylerwalls) [inline djangoproject/templates/releases/download.html]: optional: just "support periods." ("release series" being its own plural is potentially distracting?)

File: releases/tests.py [hunk 6]
```diff
@@ -817,6 +863,12 @@ def setUpTestData(cls):
                 ("rc1", datetime.date(2026, 7, 22)),
                 ("", datetime.date(2026, 8, 5)),  # final
             ],
+            "2028.0": [
+                ("a1", datetime.date(2027, 10, 1)),
+                ("b1", datetime.date(2027, 11, 1)),
+                ("rc1", datetime.date(2027, 12, 20)),
```
Comment (jacobtylerwalls) [inline releases/tests.py]: Nothing needs to change for the test, but just mentioning my understanding is that we'd do the RC much earlier in December, and grow the typical 2 week interval to final to 4.

File: releases/urls.py [hunk 7]
```diff
@@ -7,5 +7,5 @@
     re_path(
         "^([0-9a-z_.-]+)/(tarball|wheel|checksum)/$", redirect, name="download-redirect"
     ),
-    re_path(r"^(?P<series>\d{1,2}\.[0-2])/roadmap/$", roadmap, name="roadmap"),
+    re_path(r"^(?P<series>\d+\.[0-2])/roadmap/$", roadmap, name="roadmap"),
```
Comment (jacobtylerwalls) [inline releases/urls.py]: I see we will update the roadmap template in another MR.

File: releases/tests.py [hunk 8]
```diff
@@ -281,6 +319,8 @@ def test_version_tuple(self):
             ("1.8rc1", (1, 8, 0, "rc", 1)),
             ("5.2", (5, 2, 0, "final", 0)),
             ("5.2a1", (5, 2, 0, "alpha", 1)),
+            ("2028.0a1", (2028, 0, 0, "alpha", 1)),
+            ("2028.0", (2028, 0, 0, "final", 0)),
```
Comment (jacobtylerwalls) [inline releases/tests.py]: This fails
```suggestion
            ("2028", (2028, 0, 0, "final", 0)),
            ("2028.0", (2028, 0, 0, "final", 0)),
```

File: releases/management/commands/generate_release_roadmap.py [hunk 9]
```diff
@@ -111,109 +141,47 @@ def calculate_dimensions(config: dict, num_releases: int) -> int:
...
-
-    # Month labels only for the VERY FIRST set of lines
-    FIRST_YEAR_MONTH_LABELS = {
-        0: None,
-        1: "April",
-        2: "August",
-        3: "December",
-    }
+
+    month_lines = (
```
Comment (jacobtylerwalls) [inline releases/management/commands/generate_release_roadmap.py:159]: optional: We could stop drawing lines on the chart after 6.2 goes EOL?

**Review by jacobtylerwalls — COMMENTED** [review]
Marvelous, you found the right amount of changes to tackle first IMO!

File: djangoproject/templates/releases/download.html (hunk 3 above)
Comment (jacobtylerwalls) [reply djangoproject/templates/releases/download.html]: In other words, you can keep "Beginning with Django 2028.0" if you wish, but "feature releases ... YYYY.0" and "feature releases ... A.B" don't seem analogous.

Comment (sarahboyce) [reply djangoproject/templates/releases/download.html]: I think where I am slightly struggling is we have a 6.2 release being something you can pip install but not 2028, we release 2028.0 but we released 6.2 (if that makes sense)

File: releases/tests.py (hunk 8 above)
Comment (sarahboyce) [reply releases/tests.py]: Does that one need to be handled? I think we won't release `"2028"`

File: djangoproject/templates/releases/download.html (hunk 3 above)
Comment (jacobtylerwalls) [reply djangoproject/templates/releases/download.html]: I see. I didn't think about pip. I still think a reader might have my confusion, so what are your thoughts about:
```
Beginning with Django 2028.0, feature releases will use YYYY[.N]
Earlier feature releases used A.B[.C]
```

File: releases/tests.py (hunk 8 above)
Comment (jacobtylerwalls) [reply releases/tests.py]: Right, good point.

File: djangoproject/templates/releases/download.html (hunk 3 above)
Comment (carltongibson) [reply djangoproject/templates/releases/download.html]: Not sure I follow the pip point: 

```
❯ pip install 'Django==6.*'
Collecting Django==6.*
  Using cached django-6.1-py3-none-any.whl.metadata (3.9 kB)
Requirement already satisfied: asgiref>=3.9.1 in ./lib/python3.12/site-packages (from Django==6.*) (3.12.1)
Requirement already satisfied: sqlparse>=0.5.0 in ./lib/python3.12/site-packages (from Django==6.*) (0.5.5)
Using cached django-6.1-py3-none-any.whl (8.4 MB)
Installing collected packages: Django
Successfully installed Django-6.1
```

So I'm imagining the instruction to be: 

```
pip install Django==2028.*
```

Which will get you the latest patch release. 

But the feature version should be YYYY, rather than YYYY.N (where N is the patch number) I think 🤔

**Review by carltongibson — COMMENTED** [review]
This is lovely. 👍 

I think the only quibble is about not using the .0 to refer to the feature release: YYYY being equivalent to X.Y now, with YYYY.N equivalent to X.Y.Z. 

(Yesterday's release was 6.1.0, even though we only ever say "6.1")

File: djangoproject/templates/releases/download.html (hunk 3 above)
Comment (sarahboyce) [reply djangoproject/templates/releases/download.html]: I was confused by how we usually omit the .0 but have discussed the .0 in the DEP
Screenshot from 2026-08-07 09-59-47


Do we think we will have the same logic that we omit the .0 when displaying the feature release?
I think this logic is in `django.utils.version.get_main_version()` for example

Comment (carltongibson) [reply djangoproject/templates/releases/download.html]: I think that's the goal. We want to be talking about "Django 2028" or "Django 2030", where "2023.5" is for talking about the patch release.

Like, strictly "6.1" **is** "6.1.0" right? We just don't speak that way.

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (carltongibson) [reply djangoproject/templates/releases/download.html:116]: I think you mean the [6.2 row below](https://github.com/django/djangoproject.com/pull/2733/changes/BASE..126def378ee1b833b58ad4060b7331112a37cea4#diff-9ece7fd6d5728200f2be6ed04905593d65933c02c891648158db11232c915f31R133) 🤔

I'd likely say this is OK as it is: 6.2 **will** get its final _mainstream release_ at the beginning of December. At that point it's still covered for security releases, and the new version is just about to arrive. Nobody needs anything from a beginning of Jan release beyond that. 🤷

File: releases/urls.py (hunk 7 above)
Comment (sarahboyce) [reply releases/urls.py]: I updated this a little bit

File: releases/management/commands/generate_release_roadmap.py (hunk 9 above)
Comment (sarahboyce) [reply releases/management/commands/generate_release_roadmap.py:159]: I think it looks a little nicer with consistent lines but we can simplify the graph quite significantly once 5.2 is EOL in April 2028

File: releases/models.py [hunk 10]
```diff
@@ -308,10 +310,18 @@ def is_pre_release(self):
         """Return True if this is an alpha, beta, or rc release."""
         return self.status != "f"
 
+    @cached_property
+    def is_year_version(self):
```
Comment (jacobtylerwalls) [inline releases/models.py]: In a future MR we could add a CheckConstraint that micro == 0 if major > 999, to prevent data entry/test fixture mistakes.

File: releases/views.py [hunk 11]
```diff
@@ -32,7 +32,9 @@ def index(request):
 
 
 def roadmap(request, series):
-    major, minor = series.split(".")
+    major, separator, minor = series.partition(".")
```
Comment (jacobtylerwalls) [inline releases/views.py]: Nice. If the roadmap template is in scope now, you might adjust the RC section to not say "two weeks", maybe like:

```
if is_year_version
    will be released in January
```
or sim.

**Review by jacobtylerwalls — APPROVED** [review]
👍

**Review by carltongibson — APPROVED** [review]
I think this is great. (No doubt we'll make more tweaks, but this is the most of it for sure!) 👍

File: releases/management/commands/generate_release_roadmap.py [hunk 12]
```diff
@@ -64,24 +53,65 @@
     "month_line_width": 1,
 }
 
+RELEASES = [
```
Comment (pauloxnet) [inline releases/management/commands/generate_release_roadmap.py:58]: Worth adding a note somewhere obvious that this needs a manual update each
January, otherwise it will silently go stale.

File: releases/models.py [hunk 13]
```diff
@@ -308,10 +310,18 @@ def is_pre_release(self):
...
+        return len(str(self.major)) == 4
```
Comment (pauloxnet) [inline releases/models.py]:
```suggestion
        return self.major >= 2028
```
Would `major >= 2028` be safer than just checking for four digits? Avoids any
weird edge cases with test fixtures using arbitrary version numbers.

**Review by pauloxnet — APPROVED** [review]
Great work @sarahboyce ! Left a few small and non blocking comments below.

---

P.S. @carltongibson one thing I noticed looking at the SVG preview: between January and April 2030 we'll briefly have 4 versions supported at the same time: 6.2, 2028, 2029 and 2030. That's could means a few extra months of overhead for the fellows, the security team and third-party maintainers.

Since 6.2 hasn't been released yet, would it make sense to shorten its support window to December 2029 instead of April 2030? No one is relying on that date yet so nothing would be broken, and we'd stay at three concurrent versions throughout the transition, which feels cleaner.

**Comment by carltongibson** [comment]
@pauloxnet The DEP's plan is to leave 6.2 support commitment unchanged. I think that's right. (The correct thing to do.) 

I'll raise it with the Fellows though to discuss!

File: releases/models.py (hunk 13 above)
Comment (sarahboyce) [reply releases/models.py]: Good idea :+1: 
I also like how this documents which release was the first release of this format

File: releases/management/commands/generate_release_roadmap.py (hunk 12 above)
Comment (sarahboyce) [reply releases/management/commands/generate_release_roadmap.py:58]: I've updated the file's docstring and I've now added a couple of TODOs :+1: 
We will be able to simplify the logic and make this management command generate the SVG dynamically again once we're out of the transition period and we have a regular rhythm again

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (nessita) [reply djangoproject/templates/releases/download.html:116]: Isn't what Jacob is saying that even when we fully switch to YYYY versions (i.e. no 6.2 at play), there will be a month without any version with mainstream support? i.e. 2028 is EOM in December 2028 and 2029 is released ~45 days later (mid Jan 2029)? Feels like perhaps YYYY maintenance support should be dropped on Jan next year (ie cover the whole December too)?

Comment (jacobtylerwalls) [reply djangoproject/templates/releases/download.html:116]: Yes. However, I understand Carlton's view to be that it's not an issue if there's a month with no regression fixes. 8 months of regression fixes being enough, no reason to stretch the window to 9 months over the holidays. I can dig it. I just thought the diagram looked weird without it, and if we _close the gap_ it won't _look like an oversight_.

We're in what would have been the ninth month of mainstream support for 6.0, and nothing has come up so far 🤷

Comment (jacobtylerwalls) [reply djangoproject/templates/releases/download.html:116]: > (i.e. no 6.2 at play)

Sorry, now that I re-read your question, no, I did mean the one-time-only gap in from December 2027 to January 2028 (the gap on the graph).

> Feels like perhaps YYYY maintenance support should be dropped on Jan next year (ie cover the whole December too)?

I think Sarah's changes already show that.

**Comment by pauloxnet** [comment]
BTW reviewing this PR I had some idea to improve the release graph that I wrote down in the issue #2758

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (nessita) [reply djangoproject/templates/releases/download.html:116]: Well, it depends on how you read it. The download image does show the yearly releases "ending" at the end of the year, right(ish) when the next one starts. Given our current semantics, I think we should say YYYY ends mainstream support in "Jan YYYY+1". For me that is clearer to reason about (and that speech would match the graph).

File: djangoproject/templates/releases/download.html
Comment (nessita) [inline djangoproject/templates/releases/download.html]: This component will need update too, `slice:":3"` gives "202" for a version like `2028a1`, so this heading reads "the alpha for 202" and the `{% url 'roadmap' %}` below raises `NoReverseMatch`, taking down the whole download page during the pre-release cycle. I will push a fix using `preview.feature_version` instead.

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (jacobtylerwalls) [reply djangoproject/templates/releases/download.html:116]: I think we should at least update the blog post to show EOLs in January, not December.

Comment (nessita) [reply djangoproject/templates/releases/download.html:116]: Yeap agreed!

File: releases/models.py (hunk 10 above)
Comment (nessita) [inline releases/models.py]: I think we should name this after the scheme, not the cadence. "Calendar version" is the term the DEP and the blog post use, and it stays accurate if we ever move to two feature releases a year (the DEP leaves `YYYY.M.N` open), whereas "year version" would quietly become wrong while the check stayed correct:

```suggestion
    def is_calendar_version(self):
```

(I'll push this and other changes in a new commit for your evaluation.)

File: releases/models.py [hunk 14]
```diff
@@ -263,7 +263,7 @@ def version_tuple(self):
         """Return a tuple in the format of django.VERSION."""
         version = self.version.replace("-", "").replace("_", "")
         version = list(get_loose_version_tuple(version))
-        if len(version) == 2:
+        while len(version) < 3:
```
Comment (nessita) [inline releases/models.py]: I think this method needs more re-work since as is it doesn't survive the calendar pre-releases. `version_tuple` identifies the components by position, which assumes two numbers before the status letter, so `2028a1` parses as `(2028, 'a', 1, 'final', 0)` and saving it fails with "Field 'minor' expected a number but got 'a'".

In my optional commit with suggestions, I've moved the parsing to `get_version_tuple()` in `releases/utils.py`, which splits the leading numbers from the status and its iteration instead of patching
the list by index, and pads the numbers to three. Same results for everything in the A.B[.C] form, and `2028a1` now gives `(2028, 0, 0, 'alpha', 1)`. Tests cover the round trip between every version string we publish and its tuple, in both schemes.

File: releases/templatetags/release_notes.py
Comment (nessita) [inline releases/templatetags/release_notes.py]: The following may be a little of scope creep but as this stands, it does not work for the new calendar versions: it takes the first two numeric components, so `2028.3` links to `/en/2028.3/releases/2028.3/` and `2028a1` to `/en/2028.0/releases/2028.0/`, neither of which exists. What it actually wants is
the feature version, which is the same thing `Release.feature_version` needs, so I've pulled that into `get_feature_version()` in `releases/utils.py` and used it here. Patch releases now point at `/en/2028/releases/2028.3/` and pre-releases at `/en/2028/releases/2028/`, matching what we do for 5.2/6.0/6.1 etc.

File: releases/urls.py [hunk 15]
```diff
@@ -7,5 +7,5 @@
     re_path(
         "^([0-9a-z_.-]+)/(tarball|wheel|checksum)/$", redirect, name="download-redirect"
     ),
-    re_path(r"^(?P<series>\d{1,2}\.[0-2])/roadmap/$", roadmap, name="roadmap"),
+    re_path(r"^(?P<series>(?:\d+\.[0-2]|\d{4}))/roadmap/$", roadmap, name="roadmap"),
```
Comment (nessita) [inline releases/urls.py]: Making the left part of the regex `\d+` also matches `2028.0`, so `/download/2028.0/roadmap/` becomes a second live URL for the same page as `/download/2028/roadmap/`. Since the calendar series are covered by the `\d{4}` branch, I've put the A.B branch back to `\d{1,2}`.

(And we dont want to support YYYY.N links for roadmaps)

**Review by nessita — COMMENTED** [review]
I've pushed some changes as a separated commit, and left inline comments on most of
them. Two things drove the set as a whole.

First, the download page keeps asking the same question in different places: given a version string, which feature release does it belong to? Every place answered it on its own, by slicing the string or indexing the parsed components, and each of those breaks differently on `YYYY[.N]`. I've pulled the shared pieces into `releases/utils.py` (`get_version_tuple()`, `get_feature_version()`, `get_main_version()`, `is_calendar_version()`), so the model, the template tag and the views now give the same answer. That's what makes the individual fixes below one-liners.

Second, once the site can hold a 2028 release, the code around the page has to cope with one. I created a 2028 release locally to check the tables and hit the rest. The one I haven't commented on inline is artifact naming: `get_version()` drops a trailing zero from the third component, so `(2028, 0, 0, 'final', 0)` renders as "2028.0", resulting in the admin rejecting `Django-2028.tar.gz` and the checksum lands at `pgp/Django-2028.0.checksum.txt`. `get_main_version()` now picks the component to drop based on the scheme, and tests cover the round trip between every version string we publish and its tuple.

I've also respelled the pre-releases in the test data from `2028.0a1` to `2028a1`: `YYYY` is to `YYYY.N` what `A.B` is to `A.B.C`, and we ship 6.1a1 rather than 6.1.0a1. The parser needed fixing for that to be storable at all. (I also discussed this in slack with the SC.)

None of this is live until a calendar-versioned release exists, but it seemed better to land it with the change that introduces them than to leave the site in a state where creating one breaks pages.

File: releases/tests.py [hunk 16]
```diff
@@ -325,14 +336,28 @@ def test_version_tuple(self):
...
                 release = Release.objects.create(version=version)
                 self.assertEqual(release.version_tuple, expected)
 
+    def test_invalid_version(self):
+        # Malformed versions are rejected when saving rather than when parsing,
+        # as get_version_tuple() passes unknown statuses and extra numeric
+        # components through.
+        cases = [
+            ("5.2dev1", KeyError, "'dev'"),
+            ("1.2.3.4", ValueError, "too many values to unpack"),
```
Comment (jacobtylerwalls) [inline releases/tests.py]: I think we should relax the specificity of these assertions.

File: releases/tests.py [hunk 17]
```diff
@@ -512,6 +626,19 @@ def test_upload_to_artifact(self):
                 "django-5.2b2-py3-none.whl",
                 "releases/5.2/django-5.2b2-py3-none.whl",
             ),
+            # All the releases in a year series share a directory.
+            ("2028", "django-2028.tar.gz", "releases/2028/django-2028.tar.gz"),
+            ("2028.1", "django-2028.1.tar.gz", "releases/2028/django-2028.1.tar.gz"),
```
Comment (jacobtylerwalls) [inline releases/tests.py]: My inclination would have been to explore the release/artifact/documentrelease changes in a separate stream of MRs, and possibly after debugging https://github.com/django/djangoproject.com/issues/2373 first.

If we take this now, I'll want to do some manual testing myself before the next security release, whichever month that turns out to be.

File: releases/models.py (hunk 10 above)
Comment (sarahboyce) [reply releases/models.py]: This is a better name :+1:

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (sarahboyce) [reply djangoproject/templates/releases/download.html:116]: > I think we should at least update the blog post to show EOLs in January, not December.

I agree

I feel the graph is ok. It shows that 2028 mainstream support ends when 2029 is released (which is true) that the year line means mid January rather than the first of January I feel we can live with given we have a table below with specific dates and we already labeled the line before with "December" and it is clearly after that?

File: releases/tests.py [hunk 18]
```diff
@@ -871,3 +1120,31 @@ def test_links_to_contributing_and_release_process_present(self):
         response = self.client.get(url)
         self.assertContains(response, 'en/dev/internals/contributing/"')
         self.assertContains(response, 'en/dev/internals/release-process/"')
+
+    def test_release_candidate_description_year_version(self):
```
Comment (sarahboyce) [inline releases/tests.py]:
```suggestion
    def test_release_candidate_description_calendar_version(self):
```
I think as `is_year_version` got renamed to `is_calendar_version` we should update the test names where it says "year" to "calendar"

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (carltongibson) [reply djangoproject/templates/releases/download.html:116]: >  Carlton's view to be that it's not an issue if there's a month with no regression fixes. 8 months of regression fixes being enough, no reason to stretch the window to 9 months over the holidays.

>  I think we should say YYYY ends mainstream support in "Jan YYYY+1"

OK, so the DEP explicitly went for _strictly 35 months_ in order to say that mainstream support ends at the end of the third year, not rolling into the January. 

I don't think the extra month matters. 

Except nobody needs that **13th** (Jan to Dec inclusive is 12 already) maintenance release in the first week of January, and having been in the role of making that January release, it's a much more human policy to just say, you know what folks, we're not making that commitment over the new year holiday period. Security (and major bugs) is still covered, and the brand new version is two weeks away: you can cope. 

When I was Fellow, I would have much preferred a stance like that, and I assume that's true for you all, and all future Fellows too. No? 🙂

I think we should go with the DEP as written, both for this, and for the 6.2 timeline: Let it be out of mainstream support in the December: the new version is coming January. (And again, it's still in extended support, so security releases etc still happen)

WDTY?

**Comment by carltongibson** [comment]
@sarahboyce @nessita @jacobtylerwalls I'm trying to get clear on what's outstanding, dividing between the narrow topic of the PR "Update the download page to show the new roadmap" and the other bits which are "Make sure the site is 100% 2028 ready"? (Some of which are https://code.djangoproject.com/ticket/37271 &co related)

File: releases/tests.py (hunk 18 above)
Comment (nessita) [reply releases/tests.py]: Excellent point, I addressed this and pushed fixes in my commit.

File: releases/tests.py (hunk 17 above)
Comment (nessita) [reply releases/tests.py]: I was feeling a bit lost by not being able to test the whole cycle (release artifact -> download page showing it -> roadmap page) so I needed these fixes and in my view these all are "download page components". This is how I found the "download preview" issue.

I can try to separate them in a second isolated commit if you prefer?

File: releases/tests.py (hunk 16 above)
Comment (nessita) [reply releases/tests.py]: Done.

**Comment by sarahboyce** [comment]
> That's could means a few extra months of overhead for the fellows, the security team and third-party maintainers.
>
> 
> Since 6.2 hasn't been released yet, would it make sense to shorten its support window to December 2029 instead of April 2030? No one is relying on that date yet so nothing would be broken, and we'd stay at three concurrent versions throughout the transition, which feels cleaner.

@pauloxnet thank you for pointing this out :star: 
We discussed this among the Fellows, and we're comfortable supporting more releases than usual during this period. We feel that keeping the full three-year support period for the 6.2 LTS provides a smoother transition and migration experience for our user base, while also giving us one less deviation from the norm to communicate :+1:

File: djangoproject/templates/releases/download.html (hunk 1 above)
Comment (nessita) [reply djangoproject/templates/releases/download.html:116]: We discussed this among the Fellows, and our preference is to move the support transitions to January, tied to the new feature release.

The main reason is consistency and clarity: keeping the transitions tied to feature releases preserves the existing release continuum. So the release of Django YYYY marks the end of mainstream support for YYYY-1 and the end of extended support (EOL) for YYYY-3.

We understand the motivation of reducing January releaser work, but from our perspective the practical saving is limited, since December/early January may still involve security releases and final-release preparation. On balance, we prefer the cleaner January boundary.

File: releases/tests.py [hunk 19]
```diff
@@ -456,6 +626,19 @@ def test_upload_to_artifact(self):
                 "django-5.2b2-py3-none.whl",
                 "releases/5.2/django-5.2b2-py3-none.whl",
             ),
+            # All the releases in a calendar series share a directory.
+            ("2028", "django-2028.tar.gz", "releases/2028/django-2028.tar.gz"),
```
Comment (jacobtylerwalls) [inline releases/tests.py:630]: Since we're tackling the artifact logic in this PR now, I think we need negative test cases to catch cases like this:
```py
        # Invalid cases.
        for version, filename in [
            ("2028", "django-2028.1.tar.gz"),
        ]:
            with self.assertRaises(ValidationError):
                upload_to_artifact(Release(version=version), filename=filename),
```
```py
FAIL: test_upload_to_artifact (releases.tests.ReleaseUploadToTestCase.test_upload_to_artifact)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/djangoproject.com/releases/tests.py", line 653, in test_upload_to_artifact
    with self.assertRaises(ValidationError):
         ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
AssertionError: ValidationError not raised
```

File: releases/models.py [hunk 20]
```diff
@@ -308,10 +299,19 @@ def is_pre_release(self):
...
         return self.status != "f"
 
+    @cached_property
+    def is_calendar_version(self):
+        """Return True if this release is calendar versioned (YYYY[.N])."""
+        return self.major >= FIRST_CALENDAR_VERSION_YEAR
+
     @cached_property
     def is_dot_zero(self):
         """Return True if this is a final X.Y.0 release."""
```
Comment (jacobtylerwalls) [inline releases/models.py]: May as well update the docstring.

**Review by jacobtylerwalls — COMMENTED** [review]
I'm comfortable with the increased scope: the changes in releases/models.py have a smaller blast radius than I expected. One requested test & fix and I think this is g2g.

File: releases/tests.py (hunk 17 above)
Comment (jacobtylerwalls) [reply releases/tests.py]: Thanks for offering, no need. After taking a another look today, the changes have a smaller impact than I expected, so my comment was a little premature.

**Review by nessita — APPROVED** [review]
This looks great and it's very exciting!

I will merge and deploy tomorrow if no one objects.
